docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,507 @@
+---
+title: ClusterClass and managed topologies
+authors:
+  - "@srm09"
+  - "@vincepri"
+  - "@fabriziopandini"
+  - "@CecileRobertMichon"
+reviewers:
+  - "@vincepri"
+  - "@fabriziopandini"
+  - "@CecileRobertMichon"
+  - "@enxebre"
+  - "@schrej"
+creation-date: 2021-05-26
+status: provisional
+replaces:
+  - [Proposal Google Doc](https://docs.google.com/document/d/1lwxgBK3Q7zmNkOSFqzTGmrSys_vinkwubwgoyqSRAbI/edit#)
+---
+
+# ClusterClass and Managed Topologies
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Glossary](#glossary)
+      - [ClusterClass](#clusterclass)
+      - [Managed topology](#managed-topology)
+      - [Worker class](#worker-class)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+      - [Goals](#goals)
+      - [Prospective future Work](#prospective-future-work)
+  - [Proposal](#proposal)
+      - [User Stories](#user-stories)
+        - [Story 1 - Use ClusterClass to easily stamp clusters](#story-1---use-clusterclass-to-easily-stamp-clusters)
+        - [Story 2 - Easier UX for kubernetes version upgrades](#story-2---easier-ux-for-kubernetes-version-upgrades)
+        - [Story 3 - Easier UX for scaling workers nodes](#story-3---easier-ux-for-scaling-workers-nodes)
+      - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+        - [New API types](#new-api-types)
+            - [ClusterClass](#clusterclass-1)
+        - [Modification to existing API Types](#modification-to-existing-api-types)
+            - [Cluster](#cluster)
+        - [Validations](#validations)
+            - [ClusterClass](#clusterclass-2)
+            - [Cluster](#cluster-1)
+        - [Behaviors](#behaviors)
+            - [Create a new Cluster using ClusterClass object](#create-a-new-cluster-using-clusterclass-object)
+            - [Update an existing Cluster using ClusterClass](#update-an-existing-cluster-using-clusterclass)
+        - [Provider implementation](#provider-implementation)
+            - [For infrastructure providers](#for-infrastructure-providers)
+            - [For Control plane providers](#for-control-plane-providers)
+      - [Risks and Mitigations](#risks-and-mitigations)
+  - [Alternatives](#alternatives)
+  - [Upgrade Strategy](#upgrade-strategy)
+  - [Additional Details](#additional-details)
+      - [Test Plan [optional]](#test-plan-optional)
+      - [Graduation Criteria [optional]](#graduation-criteria-optional)
+      - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+  - [Implementation History](#implementation-history)
+
+## Glossary
+
+### ClusterClass
+A collection of templates that define a topology (control plane and machine deployments) to be used in one or more clusters.
+
+### Managed topology
```

Comment:
Found a similar themed comment: https://github.com/kubernetes-sigs/cluster-api/pull/4678#discussion_r644375289

Reply:
The term `Managed` explicitly emphasizes that the topology bits are managed by the Controller. On the other hand, clusters without CC have their topology bits managed by the operators. The term topology could be used to describe the cluster structure but the prefix `Managed` just makes the intent explicit.

The only other modifier that I can think of is **Supervised** but I still feel it does not correctly describe the intent.

Reply:
To follow the same logic, do we call a `Machine` a `ManagedMachine` because the bits are managed by the Controller?

Reply:
This particular scenario is different wherein a different set of CRDs are managed by a controller which isn't primarily responsible for those ones. Hence the extra emphasis on the managed aspect.

Reply:
This may have some deeper meaning for someone really close to the project, but to an outsider, this will be a difference without a distinction. 

Again, will there be other types of Topologies? Will there be UnmanagedTopologies?

Reply:
There is no concept of `UnmanagedTopology` in the works or planned in the future. I am ok to drop the **Managed** prefix in the name and add the managed aspect to the definition of the topology itself.